Index: 1.0-r5/serial-getty@.service
===================================================================
--- 1.0-r5.orig/serial-getty@.service
+++ 1.0-r5/serial-getty@.service
@@ -30,7 +30,7 @@ Before=rescue.service
 
 [Service]
 Environment="TERM=@TERM@"
-ExecStart=-/sbin/agetty -8 -L %I @BAUDRATE@ $TERM
+ExecStart=-/sbin/agetty -8 -L --autologin root %I @BAUDRATE@ $TERM
 Type=idle
 Restart=always
 UtmpIdentifier=%I
