Index: git/units/serial-getty@.service.in
===================================================================
--- git.orig/units/serial-getty@.service.in
+++ git/units/serial-getty@.service.in
@@ -32,7 +32,7 @@ Before=rescue.service
 [Service]
 # The '-o' option value tells agetty to replace 'login' arguments with '--' for
 # safety, and then the entered username.
-ExecStart=-/sbin/agetty -o '-- \\u' --noreset --noclear --keep-baud 115200,57600,38400,9600 - ${TERM}
+ExecStart=-/sbin/agetty -a root -o '-- \\u' --noreset --noclear --keep-baud 115200,57600,38400,9600 - ${TERM}
 Type=idle
 Restart=always
 UtmpIdentifier=%I
