Index: git/src/shared/machine-id-setup.c
===================================================================
--- git.orig/src/shared/machine-id-setup.c
+++ git/src/shared/machine-id-setup.c
@@ -3,6 +3,7 @@
 #include <fcntl.h>
 #include <sched.h>
 #include <sys/mount.h>
+#include <sys/stat.h>
 #include <unistd.h>
 
 #include "sd-daemon.h"
@@ -24,6 +25,7 @@
 #include "namespace-util.h"
 #include "path-util.h"
 #include "process-util.h"
+#include "sha256.h"
 #include "stat-util.h"
 #include "string-util.h"
 #include "sync-util.h"
@@ -48,6 +50,91 @@ static int acquire_machine_id_from_crede
         return 0;
 }
 
+static int read_file_contents(const char *path, char **buf_out, size_t *size_out) {
+        int r;
+        ssize_t nr;
+        struct stat statbuf;
+        _cleanup_close_ int fd = -1;
+        char *buf = NULL;
+
+        assert(path);
+        assert(buf_out);
+        assert(size_out);
+
+        fd = open(path, O_RDONLY|O_CLOEXEC|O_NOCTTY);
+        if (fd < 0) {
+                if (errno != ENOENT)
+                        return log_error_errno(-1, "Failed to open %s", path);
+                return -1; /* no error log if it doesn't exist */
+        }
+
+        if (fstat(fd, &statbuf) < 0)
+                return log_error_errno(-1, "Failed to stat %s", path);
+
+        if (statbuf.st_size == 0)
+                return log_error_errno(-1, "%s has size zero", path);
+
+        buf = malloc(statbuf.st_size);
+        if (buf == NULL)
+                return log_error_errno(-1, "Failed to malloc %zu bytes to read %s",
+                                       (size_t)statbuf.st_size, path);
+
+        nr = read(fd, buf, statbuf.st_size);
+        if (nr < 0) {
+                free(buf);
+                return log_error_errno(nr, "Failed to read %s", path);
+        } else if (nr < statbuf.st_size) {
+                free(buf);
+                log_error("Failed to read %s, got only %zd out of %zu bytes",
+                          path, nr, (size_t)statbuf.st_size);
+                return -1;
+        }
+
+        *buf_out = buf;
+        *size_out = statbuf.st_size;
+        return 0;
+}
+
+static int acquire_rpi_machine_id(sd_id128_t *ret) {
+        /* Generate a machine ID based on SHA256 of the RPi's serial
+         * number in /proc/device-tree/serial-number.
+         * Useful for keeping persistent machine-id (journal) on a read-only rootfs
+         * without an initramfs pre-mounting /etc/machine-id */
+        _cleanup_free_ char *buf = NULL;
+        struct sha256_ctx hctx;
+        uint8_t hbuf[SHA256_DIGEST_SIZE];
+        size_t buf_size = 0;
+        int r;
+
+        /* read the model name from DT to verify we're actually a Pi */
+        r = read_file_contents("/proc/device-tree/model", &buf, &buf_size);
+        if (r < 0)
+                return -1;
+        if (strncmp(buf, "Raspberry Pi", 12))
+                return -1; /* successfully read model info but we're not a Pi */
+        free(buf);
+        buf = NULL;
+
+        /* read the serial number */
+        r = read_file_contents("/proc/device-tree/serial-number", &buf, &buf_size);
+        if (r < 0)
+                return -1;
+
+        sha256_init_ctx(&hctx);
+        sha256_process_bytes(buf, buf_size, &hctx);
+        sha256_finish_ctx(&hctx, hbuf);
+
+        /* sha256 is twice as big as we need for a machine-id, just use the first half */
+        memcpy(ret, hbuf, sizeof(*ret));
+
+        /* make it a v4 UUID just like sd_id128_randomize, copied from sd-id128.c */
+        ret->bytes[6] = (ret->bytes[6] & 0x0F) | 0x40;
+        ret->bytes[8] = (ret->bytes[8] & 0x3F) | 0x80;
+
+        log_info("Generated machine-id from Raspberry Pi serial number");
+        return 0;
+}
+
 static int acquire_machine_id(const char *root, sd_id128_t *ret) {
         _cleanup_close_ int fd = -EBADF;
         int r;
@@ -101,6 +188,10 @@ static int acquire_machine_id(const char
                 }
         }
 
+        /* If that didn't work, check if we're a Raspberry Pi and use its serial number */
+        if (acquire_rpi_machine_id(ret) == 0)
+                return 0;
+
         /* If that didn't work, generate a random machine id */
         r = sd_id128_randomize(ret);
         if (r < 0)
