Index: git/tools/symbol_why.py
===================================================================
--- git.orig/tools/symbol_why.py
+++ git/tools/symbol_why.py
@@ -1028,6 +1028,9 @@ if do_blame or show_only_mismatch:
             blame_string = "## .config: {} :".format(cnt)
             o,val = split_option( line )
             if o:
+                # HACK! Ignore CONFIG_LOCALVERSION during do_kernel_configcheck. There's no nice way to allow this mismatch
+                if o == "CONFIG_LOCALVERSION":
+                    continue
                 last_val = None
                 try:
                     if option_dict[o]:
