From 9250efc71cc7613ffc9a1941d7b54df128fb741f Mon Sep 17 00:00:00 2001
From: Allen Wild <allenwild93@gmail.com>
Date: Fri, 10 Mar 2017 21:08:19 -0500
Subject: [PATCH 1/4] Makefile: allow setting INSTALL_HDR_PATH through the
 environment

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index a67c5179052a..68816b981928 100644
--- a/Makefile
+++ b/Makefile
@@ -1141,7 +1141,7 @@ headerdep:
 # Kernel headers
 
 #Default location for installed headers
-export INSTALL_HDR_PATH = $(objtree)/usr
+export INSTALL_HDR_PATH ?= $(objtree)/usr
 
 # If we do an all arch process set dst to include/arch-$(hdr-arch)
 hdr-dst = $(if $(KBUILD_HEADERS), dst=include/arch-$(hdr-arch), dst=include)
-- 
2.15.1

