From ef13494f789f3d8f809d91c45a340be6d99267f8 Mon Sep 17 00:00:00 2001
From: Allen Wild <allenwild93@gmail.com>
Date: Sat, 20 Jan 2018 00:47:27 -0500
Subject: [PATCH] crossguid: fetch tarball with yocto

---
 tools/depends/target/crossguid/Makefile | 12 ------------
 1 file changed, 12 deletions(-)

diff --git a/tools/depends/target/crossguid/Makefile b/tools/depends/target/crossguid/Makefile
index 487ab2a617..0b806d11bd 100644
--- a/tools/depends/target/crossguid/Makefile
+++ b/tools/depends/target/crossguid/Makefile
@@ -41,18 +41,6 @@ all: .installed-$(PLATFORM) $(PREFIX)/lib/lib$(LIBNAME).a
 $(PREFIX)/lib/lib$(LIBNAME).a:
 	@make .installed-$(PLATFORM)
 
-$(TARBALLS_LOCATION)/$(ARCHIVE):
-	cd $(TARBALLS_LOCATION); $(RETRIEVE_TOOL) $(RETRIEVE_TOOL_FLAGS) $(BASE_URL)/$(ARCHIVE)
-
-$(PLATFORM): $(TARBALLS_LOCATION)/$(ARCHIVE) $(DEPS)
-ifeq ($(PREFIX),)
-	@echo
-	@echo "ERROR: please set PREFIX to the kodi install path e.g. make PREFIX=/usr/local"
-	@exit 1
-endif
-	rm -rf $(PLATFORM); mkdir -p $(PLATFORM)
-	cd $(PLATFORM); $(ARCHIVE_TOOL) $(ARCHIVE_TOOL_FLAGS) $(TARBALLS_LOCATION)/$(ARCHIVE)
-
 $(LIBDYLIB): $(PLATFORM)
 	cd $(PLATFORM); $(CXX) -c guid.cpp -o guid.o $(CXXFLAGS) -D$(GUID_PLATFORM_DEFINE)
 	cd $(PLATFORM); $(AR) rvs lib$(LIBNAME).a guid.o
-- 
2.16.0

