From 253374a18c6fad8f6b6826d77d1dc04cfa6d909a Mon Sep 17 00:00:00 2001
From: Allen Wild <allenwild93@gmail.com>
Date: Fri, 10 Mar 2017 21:08:19 -0500
Subject: [PATCH 01/11] Makefile: allow setting INSTALL_HDR_PATH through the
 environment

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 14dc275..5272698 100644
--- a/Makefile
+++ b/Makefile
@@ -1123,7 +1123,7 @@ firmware_install:
 # Kernel headers
 
 #Default location for installed headers
-export INSTALL_HDR_PATH = $(objtree)/usr
+export INSTALL_HDR_PATH ?= $(objtree)/usr
 
 # If we do an all arch process set dst to asm-$(hdr-arch)
 hdr-dst = $(if $(KBUILD_HEADERS), dst=include/asm-$(hdr-arch), dst=include/asm)
-- 
2.9.3

